Create Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md (#16869)
kubernetes/website: content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md

> This document describes how to install a single-master Kubernetes cluster v1.15 with kubeadm on CentOS, and then deploy an external OpenStack cloud provider and Cinder CSI plugin in order to use Cinder volumes as persistent volumes in Kubernetes.

Reviewer: ```suggestion
This document describes how to install a single-master Kubernetes cluster v1.15 with kubeadm on CentOS, and then deploy an external OpenStack cloud provider and Cinder CSI plugin to use Cinder volumes as persistent volumes in Kubernetes.
```

> This cluster will be running on OpenStack VMs so we'll create a few things in OpenStack for it.

Reviewer: ```suggestion
This cluster will be running on OpenStack VMs, so we'll create a few things in OpenStack for it.
```

* a project/tenant for this kubernetes cluster
* a user in this project for Kubernetes, to query node information and attach volumes etc
* a private network and subnet
* a router for this private network and connect it to a public network for floating IPs
* a security group for all Kubernetes VMs

> * a VM as master node and a few VMs as worker nodes

Reviewer: ```suggestion
* a VM as control plane node and a few VMs as worker nodes
```

The security group will have the following rules to open ports for Kubernetes.

**Master Node**

|Protocol  | Port Number | Description|
|----------|-------------|------------|
|TCP |6443|Kubernetes API Server|
|TCP|2379-2380|etcd server client API|
|TCP|10250|Kubelet API|
|TCP|10251|kube-scheduler|
|TCP|10252|kube-controller-manager|
|TCP|10255|Read-only Kubelet API|

**Worker Nodes**

|Protocol  | Port Number | Description|
|----------|-------------|------------|
|TCP|10250|Kubelet API|
|TCP|10255|Read-only Kubelet API|
|TCP|30000-32767|NodePort Services|

**CNI Ports on both master and worker nodes**

|Protocol  | Port Number | Description|
|----------|-------------|------------|

> |TCP|179|Calico BGP network|

Reviewer 2: Calico and Flannel are only referenced here. The example below uses weave. It might be worth removing or calling out that these should only be added if running them as an alternative.

|TCP|9099|Calico felix (health check)|
|UDP|8285|Flannel|
|UDP|8472|Flannel|
|TCP|6781-6784|Weave|
|UDP|6783-6784|Weave|

> The master needs at least 2 cores and 4GB RAM. After the VM is launched, verify its hostname and make sure it is the same as the node name in Nova. 

Reviewer: ```suggestion
The control plane needs at least 2 cores and 4GB RAM. After the VM is launched, verify its hostname and make sure it is the same as the node name in Nova. 
```

If the hostname is not resolvable, add it to /etc/hosts.

For example, if the VM is called master1, and it has an internal IP 192.168.1.4. Add that to /etc/hosts and set hostname to master1.
```
echo "192.168.1.4 master1" >> /etc/hosts

hostnamectl set-hostname master1
```

> Next we'll follow official documents to install docker and Kubernetes using kubeadm.

Reviewer: ```suggestion
Next, we'll follow official documents to install docker and Kubernetes using kubeadm.
```

> Install docker following the steps in https://kubernetes.io/docs/setup/production-environment/container-runtimes/

Reviewer 2: ```suggestion
Install docker following the steps from the [Container Runtime Documentation.](/docs/setup/production-environment/container-runtimes/)
```
Using an internal link like this is preferred, it will enable language specific routing to the correct docs. If you could, please update it along with the others.

> Note that it is recommend to use systemd as the cgroup driver for Kubernetes. 

Reviewer: ```suggestion
Note that it is best practice to use systemd as the cgroup driver for Kubernetes.
```

```
cat <<EOF > /etc/yum.repos.d/kubernetes.repo
[kubernetes]
name=Kubernetes
baseurl=https://packages.cloud.google.com/yum/repos/kubernetes-el7-x86_64
enabled=1
gpgcheck=1
repo_gpgcheck=1
gpgkey=https://packages.cloud.google.com/yum/doc/yum-key.gpg https://packages.cloud.google.com/yum/doc/rpm-package-key.gpg
EOF

# Set SELinux in permissive mode (effectively disabling it)
setenforce 0
sed -i 's/^SELINUX=enforcing$/SELINUX=permissive/' /etc/selinux/config

yum install -y kubelet kubeadm kubectl --disableexcludes=kubernetes

systemctl enable --now kubelet

cat <<EOF >  /etc/sysctl.d/k8s.conf
net.bridge.bridge-nf-call-ip6tables = 1
net.bridge.bridge-nf-call-iptables = 1
EOF
sysctl --system

# check if br_netfilter module is loaded
lsmod | grep br_netfilter

# if not, load it explicitly with 
modprobe br_netfilter
```

The official document about how to create a single-master cluster can be found in https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/

We'll largely follow that document but also add additional things for the cloud provider.

> To make things clearer we'll use a kubeadm-config.yml for the master.

Reviewer: ```suggestion
To make things more clear, we'll use a kubeadm-config.yml for the master.
```

In this config we specify to use an external OpenStack cloud provider, and where to find its config.

> We also enable storage API in API server's runtime config so we can use OpenStack volumes as persistent volumes in Kubernets.

Reviewer: ```suggestion
We also enable storage API in API server's runtime config so we can use OpenStack volumes as persistent volumes in Kubernetes.
```

```
apiVersion: kubeadm.k8s.io/v1beta1
kind: InitConfiguration
nodeRegistration:
  kubeletExtraArgs:
    cloud-provider: "external"
---
apiVersion: kubeadm.k8s.io/v1beta2
kind: ClusterConfiguration
kubernetesVersion: "v1.15.1"
apiServer:
  extraArgs:
    enable-admission-plugins: NodeRestriction
    runtime-config: "storage.k8s.io/v1=true"
controllerManager:
  extraArgs:
    external-cloud-volume-plugin: openstack
  extraVolumes:
  - name: "cloud-config"
    hostPath: "/etc/kubernetes/cloud-config"
    mountPath: "/etc/kubernetes/cloud-config"
    readOnly: true
    pathType: File
networking:
  serviceSubnet: "10.96.0.0/12"
  podSubnet: "10.224.0.0/16"
  dnsDomain: "cluster.local"
```

> Now we'll create cloud config, /etc/kubernetes/cloud-config, for OpenStack. 

Reviewer 2: ```suggestion
Now we'll create cloud config, `/etc/kubernetes/cloud-config`, for OpenStack. 
```
File paths should use code-styling. 
See the style guide for more info: https://kubernetes.io/docs/contribute/style/style-guide/#use-code-style-for-filenames-directories-and-paths

Note that the tenant here is the one we created for all Kubernets VMs in the beginning.
All VMs should be launched in this project/tenant.
In addition you need to create a user in this tenant for Kubernetes to do queries.

> The ca-file is the CA root certitiface for OpenStack's API endpoint https://openstack.cloud:5000/v3

Reviewer 2: ```suggestion
The ca-file is the CA root certificate for OpenStack's API endpoint, for example `https://openstack.cloud:5000/v3`
```
As far as I know, this isn't always the url. I would preface this with for example and wrap it in code style, otherwise it'll come up as a 404 link if people click it.

At the time of writing the cloud provider doesn't allow insecure connections (skip CA check).

```
[Global]
region=RegionOne
username=username
password=password
auth-url=https://openstack.cloud:5000/v3
tenant-id=14ba698c0aec4fd6b7dc8c310f664009
domain-id=default
ca-file=/etc/kubernetes/ca.pem

[LoadBalancer]
subnet-id=b4a9a292-ea48-4125-9fb2-8be2628cb7a1
floating-network-id=bc8a590a-5d65-4525-98f3-f7ef29c727d5

[BlockStorage]
bs-version=v2

[Networking]
public-network-name=public
ipv6-support-disabled=false
```

Next run kubeadm to initiate the master
```
kubeadm init --config=kubeadm-config.yml
```

> When the initialisation is finished, copy admin config to .kube

Reviewer: ```suggestion
With the initialization completed, copy admin config to .kube
```

```
  mkdir -p $HOME/.kube
  sudo cp -i /etc/kubernetes/admin.conf $HOME/.kube/config
  sudo chown $(id -u):$(id -g) $HOME/.kube/config
```

> At this stage the master node is created but not ready. All the nodes have the taint <code>node.cloudprovider.kubernetes.io/uninitialized=true:NoSchedule</code> and waiting for being initialized by cloud-controller-manager. 

Reviewer: ```suggestion
At this stage, the control plane node is created but not ready. All the nodes have the taint <code>node.cloudprovider.kubernetes.io/uninitialized=true:NoSchedule</code> and waiting for being initialized by cloud-controller-manager. 
```

          path: /usr/libexec/kubernetes/kubelet-plugins/volume/exec
          type: DirectoryOrCreate
        name: flexvolume-dir
      - hostPath:
          path: /etc/kubernetes/pki
          type: DirectoryOrCreate
        name: k8s-certs
      - hostPath:
          path: /etc/ssl/certs
          type: DirectoryOrCreate
        name: ca-certs
      - name: cloud-config-volume
        secret:
          secretName: cloud-config
      - name: ca-cert
        secret:
          secretName: openstack-ca-cert
```

When the controller manager is running, it will query OpenStack to get information about the nodes and remove the taint. In the node info you'll see the VM's UUID in OpenStack.
```
# kubectl describe no master1
Name:               master1
Roles:              master
......
Taints:             node-role.kubernetes.io/master:NoSchedule
                    node.kubernetes.io/not-ready:NoSchedule
......
sage:docker: network plugin is not ready: cni config uninitialized
......
PodCIDR:                     10.224.0.0/24
ProviderID:                  openstack:///548e3c46-2477-4ce2-968b-3de1314560a5

```
Now install your favourite CNI and the master node will become ready.

> For example, to install weave net, run this command

Reviewer 2: ```suggestion
For example, to install weave net, run this command:
```

```
kubectl apply -f "https://cloud.weave.works/k8s/net?k8s-version=$(kubectl version | base64 | tr -d '\n')"
```

Next we'll set up worker nodes.

> Firstly install docker and kubeadm in the same way as how they were installed in the master. 

Reviewer 2: ```suggestion
Firstly, install docker and kubeadm in the same way as how they were installed in the master. 
```
Missing a comma.

To join them to the cluster we need a token and ca cert hash from the output of master installation. 
If it is expired or lost we can recreate it using these commands.

```
# check if token is expired
kubeadm token list

# re-create token and show join command
kubeadm token create --print-join-command

```

Create kubeadm-config.yml for worker nodes with the above token and ca cert hash.
```
apiVersion: kubeadm.k8s.io/v1beta2
discovery:
  bootstrapToken:
    apiServerEndpoint: 192.168.1.7:6443
    token: 0c0z4p.dnafh6vnmouus569
    caCertHashes: ["sha256:fcb3e956a6880c05fc9d09714424b827f57a6fdc8afc44497180905946527adf"]
kind: JoinConfiguration
nodeRegistration:
  kubeletExtraArgs:
    cloud-provider: "external"

```

> apiServerEndpoint is the master node, token and caCertHashes can be taken from the join command printed in the output of 'kubeadm token create' command.

Reviewer: ```suggestion
apiServerEndpoint is the control plane node, token and caCertHashes can be taken from the join command printed in the output of 'kubeadm token create' command.
```

Run kubeadm and the worker nodes will be joined to the cluster.
```
kubeadm join  --config kubeadm-config.yml 
```

At this stage we'll have a working Kubernetes cluster with an external OpenStack cloud provider.
The provider tells Kubernetes about the mapping between Kubernetes nodes and OpenStack VMs.
If Kubernetes wants to attach a persistent volume to a pod, it can find out which OpenStack VM the pod is running on from the mapping, and attach the underlying OpenStack volume to the VM accordingly.

> The integration with Cinder is provided by an external Cinder CSI plugin, as described in https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/using-cinder-csi-plugin.md

Reviewer 2: ```suggestion
The integration with Cinder is provided by an external Cinder CSI plugin, as described in [the Cinder CSI documentation](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/using-cinder-csi-plugin.md)
```

It's generally better to put the link behind some text instead of providing a raw link.

> We'll perform the following steps to install the Clinder CSI plugin.

Reviewer 2: ```suggestion
We'll perform the following steps to install the Cinder CSI plugin.
```

> Firstly create a secret with CA certs for OpenStack's API endpoints. It is the same cert file as what we use in cloud provider above.

Reviewer 2: ```suggestion
Firstly, create a secret with CA certs for OpenStack's API endpoints. It is the same cert file as what we use in cloud provider above.
```

```
kubectl create secret -n kube-system generic openstack-ca-cert --from-literal=ca.pem="$(cat /etc/kubernetes/ca.pem)" --dry-run -o yaml > openstack-ca-cert.yaml
kubectl apply -f openstack-ca-cert.yaml
```
Then create RBAC resources.
```
kubectl apply -f https://raw.githubusercontent.com/kubernetes/cloud-provider-openstack/release-1.15/manifests/cinder-csi-plugin/cinder-csi-controllerplugin-rbac.yaml
kubectl apply -f https://github.com/kubernetes/cloud-provider-openstack/raw/release-1.15/manifests/cinder-csi-plugin/cinder-csi-nodeplugin-rbac.yaml
```

The Cinder CSI plugin includes a controller plugin and a node plugin.

> The controller communicates with Kubernetes APIs and Cinder APIs to create/attach/detach/delete Cinder volumes, while node plugin runs on each worker node to bind a storage device (attached volume) to a pod, and unbind it during deletion.

Reviewer 2: ```suggestion
The controller communicates with Kubernetes APIs and Cinder APIs to create/attach/detach/delete Cinder volumes. The node plugin in-turn runs on each worker node to bind a storage device (attached volume) to a pod, and unbind it during deletion.
```
Breaking apart a bit sentence.

            type: Directory
        - name: pods-cloud-data
          hostPath:
            path: /var/lib/cloud/data
            type: Directory
        - name: pods-probe-dir
          hostPath:
            path: /dev
            type: Directory
        - name: secret-cinderplugin
          secret:
            secretName: cloud-config
        - name: ca-cert
          secret:
            secretName: openstack-ca-cert

```
When they are both running, create a storage class for Cinder.

```
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: csi-sc-cinderplugin
provisioner: csi-cinderplugin
```
Then we can create a PVC with this class.
```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: myvol
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 1Gi
  storageClassName: csi-sc-cinderplugin

```

When the PVC is created, a Cinder volume is created correspondingly.
```
# kubectl get pvc
NAME    STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS          AGE
myvol   Bound    pvc-14b8bc68-6c4c-4dc6-ad79-4cb29a81faad   1Gi        RWO            csi-sc-cinderplugin   3s

```

> In OpenStack the volume will be called "*pvc*-14b8bc68-6c4c-4dc6-ad79-4cb29a81faad"

Reviewer 2: ```suggestion
In OpenStack the volume name will match the Kubernetes persistent volume generated name. In this example it would be: _pvc-14b8bc68-6c4c-4dc6-ad79-4cb29a81faad_
```

| updated_at                     | 2019-07-24T05:02:35.000000                                                                                                                                                                                                                                                                                                     |
| user_id                        | 5f6a7a06f4e3456c890130d56babf591                                                                                                                                                                                                                                                                                               |
+--------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

```

> 

Reviewer 2: I think it'd be worth adding some conclusion here, or some form of final note.
